server/server/src/billing/billing_service.rs

```diff
@@ -379,113 +514,300 @@ pub async fn stripe_webhooks(
-                event
+    Ok(public_key)
+}
+
+#[derive(Debug)]
+pub enum GooglePlayWebhookError {
+    InvalidToken,
+    CannotRetrieveData,
+    CannotDecodePubSubData(DecodeError),
+    CannotRetrieveUserInfo,
+    CannotRetrievePublicKey,
+    CannotParseTime,
+}
+
+pub async fn google_play_notification_webhooks(
+    server_state: &Arc<ServerState>, request_body: Bytes, query_parameters: HashMap<String, String>,
+) -> Result<(), ServerError<GooglePlayWebhookError>> {
+    if !constant_time_eq::constant_time_eq(
+        query_parameters
+            .get("token")
+            .ok_or(ClientError(GooglePlayWebhookError::InvalidToken))?
+            .as_bytes(),
+        server_state.config.google.pubsub_token.as_bytes(),
+    ) {
+        return Err(ClientError(GooglePlayWebhookError::InvalidToken));
+    }
+
+    info!("Parsing pubsub notification and extracting the developer notification.");
+
+    let pubsub_notif = serde_json::from_slice::<PubSubNotification>(&request_body)?;
+    let data = base64::decode(pubsub_notif.message.data)
+        .map_err(|e| ClientError(GooglePlayWebhookError::CannotDecodePubSubData(e)))?;
+
+    let notification = serde_json::from_slice::<DeveloperNotification>(&data)?;
+
+    let mut con = server_state.index_db_pool.get().await?;
+
+    if let Some(sub_notif) = notification.subscription_notification {
+        info!("Notification is for a subscription: {:?}", sub_notif);
+
+        let subscription = google_play_client::get_subscription(
+            &server_state.android_publisher,
+            &sub_notif.subscription_id,
+            &sub_notif.purchase_token,
+        )
+        .await
+        .map_err(|e| internal!("{:#?}", e))?;
+
+        let notification_type = sub_notif.notification_type();
+        if let NotificationType::SubscriptionPurchased = notification_type {
+            return Ok(());
+        }
+
+        let account_id = &subscription
+            .obfuscated_external_account_id
+            .clone()
+            .ok_or_else(|| {
+                internal!("There should be an account id attached to a purchase: {:?}", sub_notif)
+            })?;
+
+        info!("Retrieved full subscription info for notification event {:?} with an obfuscated id of {:?}", notification_type, account_id);
+
+        let public_key: PublicKey = con
+            .maybe_json_get(keys::public_key_from_gp_account_id(account_id))
+            .await?
+            .ok_or_else(|| {
+                internal!("There is no public_key related to this account_id: {:?}", account_id)
+            })?;
+
+        info!("Updating subscription history to match new subscription state.");
+
+        let tx_result = tx!(&mut con, pipe, &[keys::subscription_profile(&public_key)], {
+            match lock_billing_workflow(
+                &public_key,
+                &mut con,
+                server_state.config.stripe.millis_between_user_payment_flows,
```

Comment:
It's weird that you're going into stripe's config to get this, you can organize such that there's a billing config and the common things are up there.